src/Task/TalkNPC.pm

```diff
@@ -318,28 +444,110 @@ sub iterate {
 				}
 				if (grep(defined, @bulkitemlist)) {
 					$messageSender->sendBuyBulk(\@bulkitemlist);
-					$ai_v{npc_talk}{talk} = 'close' if !$self->{steps}[0];
+					$ai_v{'npc_talk'}{'talk'} = 'close' if !$self->{steps}[0];
 				}
 				# We give some time to get inventory_item_added packet from server.
 				# And skip this itteration.
-				$ai_v{npc_talk}{time} = time + 0.2;
+				$ai_v{'npc_talk'}{'time'} = time + 0.2;
 				$self->{time} = time + 0.2;
 				return;
+				
+			# Click the cancel button in a shop.
+			} elsif ($step =~ /^e$/i) {
+				$ai_v{'npc_talk'}{'talk'} = 'close';
+				$self->conversation_end;
+				
+			# Wrong sequence
+			} else {
+				$self->set_error_and_try_to_cancel(WRONG_NPC_INSTRUCTIONS, T("This npc requires the buy or cancel button to be pressed, but the given instructions don't match that."));
 			}
+		
+		} elsif ( $current_talk_step eq 'sell' ) {
+			$self->conversation_end;
 
-		} elsif ( $step =~ /^s/i ) {
-			# Get the sell list in a shop.
-			$messageSender->sendNPCBuySellList($talk{ID}, 1);
-
-		} elsif ( $step =~ /^e/i ) {
-			# ? Pretend like the conversation was stopped by the NPC?
-			$ai_v{npc_talk}{talk} = 'close';
+		} else {
+			$self->set_error_and_try_to_cancel(WRONG_NPC_INSTRUCTIONS, T("According to the given NPC instructions, a npc conversation code " .
+				"should be used (".$step."), but it doesn't exist."));
 		}
-
+		
+		$self->{wait_for_answer} = 1;
 		shift @{$self->{steps}};
+	
+	# After a 'npc_talk_done' hook we must always send a 'npc_talk_cancel' after a timeout
+	} elsif ($self->{stage} == AFTER_NPC_CLOSE) {
+		return unless (timeOut($self->{time}, 1));
+		#Now 'n' step is totally unnecessary as we always send it but this must be done for backwards compatibility
+		if ( $self->{steps}[0] =~ /^n/i ) {
+			shift(@{$self->{steps}});
+		}
+		$self->{time} = time;
+		$self->{stage} = AFTER_NPC_CANCEL;
+		$messageSender->sendTalkCancel($self->{ID});
```

Comment:
That needs confirmation, I tested on bRO and RMS renewal test server, it is up for changes.